### ISSUE TITLE
`rustc --crate-type staticlib --emit asm` on macOS silently omits code

### ISSUE THREAD

**Issue by ob** [rust-lang/rust#44359]
I was playing with a trivial rust program:

```rust
#![feature(inclusive_range_syntax)]

#![no_mangle]
pub extern "C" fn isprime(n: u64) -> bool {
    for i in 2...n/2 {
        if n % i == 0 {
            return false;
        }
    }
    return true;
}
```

and compiling it like so

```
$ rustc -O --emit asm --crate-type staticlib isprime.rs
$ echo $?
0
$ cat isprime.s
	.section	__TEXT,__text,regular,pure_instructions

.subsections_via_symbols
```

I was very surprised that it produced no output. However, if I do

```
$ rustc -O --emit asm --crate-type lib isprime.rs
```
I get the expected output (except the name of the function is still mangled).

I _think_ this is because macOS doesn't support static libraries? It seems it should be an error if that is the case.

This is with

```
$ rustc --version
rustc 1.22.0-nightly (088216fb9 2017-09-04)
$ sw_vers
ProductName:	Mac OS X
ProductVersion:	10.12.6
BuildVersion:	16G29
```

But I tried stable as well.

**Comment by alexcrichton** [comment]
Thanks for the report! This is currently expected behavior, though, albeit a bad error message. The problem is you're using `#![...]` instead of `#[..]` for the `no_mangle` attribute, so `isprime` isn't getting to the output because it's not an exported symbol.

**Comment by ob** [comment]
Ah, I see. Hmm... that seems like a hard error to catch no? If `no_mangle` doesn't make sense as a `#![...]` attribute would it make sense to at least provide a warning if it's used like so?

**Comment by alexcrichton** [comment]
In theory we could give a warning that the `no_mangle` attribute applied on the crate isn't being used, but that's my best guess
